src/backend/wire/protocol.cpp

```diff
@@ -0,0 +1,671 @@
+ */
+void PacketManager::close_client() { client.sock->close_socket(); }
+
+
+void PacketManager::make_hardcoded_parameter_status(
+    ResponseBuffer &responses,
+    const std::pair<std::string, std::string>& kv) {
+  std::unique_ptr<Packet> response(new Packet());
+  response->msg_type = 'S';
+  packet_putstring(response, kv.first);
+  packet_putstring(response, kv.second);
+  responses.push_back(std::move(response));
+}
+/*
+ * process_startup_packet - Processes the startup packet
+ * 	(after the size field of the header).
+ */
+bool PacketManager::process_startup_packet(Packet* pkt,
+                                           ResponseBuffer& responses) {
+  std::string token, value;
+  std::unique_ptr<Packet> response(new Packet());
+
+  int32_t proto_version = packet_getint(pkt, sizeof(int32_t));
+
+  // Only protocol version 3 is supported
+  if (PROTO_MAJOR_VERSION(proto_version) != 3) {
+    LOG_ERROR("Protocol error: Only protocol version 3 is supported.");
+    exit(EXIT_FAILURE);
+  }
+
+  // TODO: check for more malformed cases
+  // iterate till the end
+  for (;;) {
+    // loop end case?
+    if (pkt->ptr >= pkt->len) break;
+    get_string_token(pkt, token);
+
+    // if the option database was found
+    if (token.compare("database") == 0) {
+      // loop end?
+      if (pkt->ptr >= pkt->len) break;
+      get_string_token(pkt, client.dbname);
+    } else if (token.compare(("user")) == 0) {
+      // loop end?
+      if (pkt->ptr >= pkt->len) break;
+      get_string_token(pkt, client.user);
+    } else {
+      if (pkt->ptr >= pkt->len) break;
+      get_string_token(pkt, value);
+      client.cmdline_options[token] = value;
+    }
+  }
+
+  // send auth-ok ('R')
+  response->msg_type = 'R';
+  packet_putint(response, 0, 4);
+  responses.push_back(std::move(response));
+
+  // Send the parameterStatus map ('S')
+  for(auto it = parameter_status_map.begin(); it != parameter_status_map.end(); it++) {
+    make_hardcoded_parameter_status(responses, *it);
+  }
+
+  // ready-for-query packet -> 'Z'
+  send_ready_for_query(TXN_IDLE, responses);
+  return true;
+}
+
+void PacketManager::put_row_desc(std::vector<wiredb::FieldInfoType> &rowdesc, ResponseBuffer &responses) {
+  if (!rowdesc.size())
+    return;
+
+  LOG_INFO("Put RowDescription");
+  std::unique_ptr<Packet> pkt(new Packet());
+  pkt->msg_type = 'T';
+  packet_putint(pkt, rowdesc.size(), 2);
+
+  for (auto col : rowdesc) {
+    LOG_INFO("column name: %s", std::get<0>(col).c_str());
+    packet_putstring(pkt, std::get<0>(col));
+    // TODO: Table Oid (int32)
+    packet_putint(pkt, 0, 4);
+    // TODO: Attr id of column (int16)
+    packet_putint(pkt, 0, 2);
+    // Field data type (int32)
+    packet_putint(pkt, std::get<1>(col), 4);
+    // Data type size (int16)
+    packet_putint(pkt, std::get<2>(col), 2);
+    // Type modifier (int32)
+    packet_putint(pkt, -1, 4);
+    // Format code for text
+    packet_putint(pkt, 0, 2);
+  }
```

Comment:
I think `packet_put_int`, `packet_put_string` are better names for the functions.